cpp/src/arrow/util/bit_util.h

```diff
@@ -827,6 +838,253 @@ Status InvertBitmap(MemoryPool* pool, const uint8_t* bitmap, int64_t offset,
+        }
+        visitor(visited_words);
+      }
+      consume(whole_word_count * kBitWidth);
+    } else {
+      // leading bits from potentially incomplete words have been consumed
+
+      // word_i such that words[i][word_i] and words[i][word_i + 1] are lie entirely
+      // within the bitmap for all i
+      for (int64_t word_i = 0; word_i < whole_word_count - 1; ++word_i) {
+        for (size_t i = 0; i < N; ++i) {
+          if (offsets[i] == 0) {
+            visited_words[i] = words[i][word_i];
+          } else {
+            visited_words[i] = words[i][word_i] >> offsets[i];
+            visited_words[i] |= words[i][word_i + 1] << (kBitWidth - offsets[i]);
+          }
+        }
+        visitor(visited_words);
+      }
+      consume((whole_word_count - 1) * kBitWidth);
+
+      SafeLoadWords(bitmaps, 0, kBitWidth, false, &visited_words);
+
+      visitor(visited_words);
+      consume(kBitWidth);
+    }
+
+    // load remaining bits
+    SafeLoadWords(bitmaps, 0, bit_length, false, &visited_words);
+    visitor(visited_words);
+
+    return min_offset;
+  }
+
+  const std::shared_ptr<Buffer>& buffer() const { return buffer_; }
+
+  /// offset of first bit relative to buffer().data()
+  int64_t offset() const { return offset_; }
+
+  /// number of bits in this Bitmap
+  int64_t length() const { return length_; }
+
+  /// string_view of all bytes which contain any bit in this Bitmap
+  util::bytes_view bytes() const {
+    auto byte_offset = offset_ / 8;
+    auto byte_count = BitUtil::CeilDiv(offset_ + length_, 8) - byte_offset;
+    return util::bytes_view(buffer_->data() + byte_offset, byte_count);
+  }
+
+ private:
+  /// string_view of all Words which contain any bit in this Bitmap
+  ///
+  /// For example, given Word=uint16_t and a bitmap spanning bits [20, 36)
+  /// words() would span bits [16, 48).
+  ///
+  /// 0       16      32     48     64
+  /// |-------|-------|------|------| (buffer)
+  ///           [       ]             (bitmap)
+  ///         |-------|------|        (returned words)
+  ///
+  /// \warning The words may contain bytes which lie outside the buffer or are
+  /// uninitialized.
+  template <typename Word>
+  View<Word> words() const {
+    auto bytes_addr = reinterpret_cast<size_t>(bytes().data());
+    auto words_addr = bytes_addr - bytes_addr % sizeof(Word);
+    auto word_byte_count =
+        BitUtil::RoundUpToPowerOf2(static_cast<int64_t>(bytes_addr + bytes().size()),
+                                   static_cast<int64_t>(sizeof(Word))) -
+        words_addr;
+    return View<Word>(reinterpret_cast<const Word*>(words_addr),
+                      word_byte_count / sizeof(Word));
+  }
+
+  /// offset of first bit relative to words<Word>().data()
+  template <typename Word>
+  int64_t word_offset() const {
+    return offset_ + 8 * (reinterpret_cast<size_t>(buffer_->data()) -
+                          reinterpret_cast<size_t>(words<Word>().data()));
```

Comment:
We can't use `size_t` here because `sizeof(size_t)` may be 4byte on 32bit OS such Windows and Linux.

```diff
-    return offset_ + 8 * (reinterpret_cast<size_t>(buffer_->data()) -
-                          reinterpret_cast<size_t>(words<Word>().data()));
+    return offset_ + 8 * (reinterpret_cast<uint64_t>(buffer_->data()) -
+                          reinterpret_cast<uint64_t>(words<Word>().data()));
```

Reply:
Well, precisely, you want to cast the `data()` pointer to an integer of the same size, so for example `size_t` or `intptr_t`.
I don't this casting a 32-bit pointer to a 64-bit integer is well-defined.

Reply:
I confirmed that `intptr_t` is also worked on 32bit Linux.

```diff
-    return offset_ + 8 * (reinterpret_cast<size_t>(buffer_->data()) -
-                          reinterpret_cast<size_t>(words<Word>().data()));
+    return offset_ + 8 * (reinterpret_cast<intptr_t>(buffer_->data()) -
+                          reinterpret_cast<intptr_t>(words<Word>().data()));
```

BTW, can we use the following here? It also works on 32bit Linux.

```diff
-    return offset_ + 8 * (reinterpret_cast<size_t>(buffer_->data()) -
-                          reinterpret_cast<size_t>(words<Word>().data()));
+    return offset_ + 8 * (buffer_->data() -
+                          reinterpret_cast<const uint8_t*>(words<Word>().data()));
```

Reply:
Yes, we can. I think the problem with `size_t` is that it's unsigned and we're promoting it to a larger signed type.
(I prefer the `intptr_t` solution aesthetically, personally)

Reply:
Huh, alright. I would've expected the reinterpet cast to unsigned to work fine. I'll use `intptr_t`